src/uma2/permission_association.rs

```diff
@@ -0,0 +1,347 @@
+    /// * `name` Name for the permission
+    /// * `description` Description for the permission
+    /// * `scopes` A list of scopes given on this resource to the user if the permission validates
+    /// * `roles` Give the permission to users in a list of roles
+    /// * `groups` Give the permission to users in a list of groups
+    /// * `clients` Give the permission to users using a specific list of clients
+    /// * `owner` Give the permission to the owner
+    /// * `logic` Positive: If the user is in the required groups/roles or using the right client, then
+    ///     give the permission to the user. Negative - the inverse
+    /// * `decision_strategy` Go through the required conditions. If it is more than one condition,
+    ///     give the permission to the user if the following conditions are met:
+    ///         - Unanimous: The default strategy if none is provided. In this case, all policies must evaluate
+    ///             to a positive decision for the final decision to be also positive.
+    ///         - Affirmative: In this case, at least one policy must evaluate to a positive decision
+    ///             in order for the final decision to be also positive.
+    ///         - Consensus: In this case, the number of positive decisions must be greater than
+    ///             the number of negative decisions. If the number of positive and negative
+    ///             decisions is the same, the final decision will be negative
+    pub async fn update_uma2_resource_permission(
+        &self,
+        id: String,
+        token: String,
+        name: String,
+        description: String,
+        scopes: Vec<String>,
+        roles: Option<Vec<String>>,
+        groups: Option<Vec<String>>,
+        clients: Option<Vec<String>>,
+        owner: Option<String>,
+        logic: Option<Uma2PermissionLogic>,
+        decision_strategy: Option<Uma2PermissionDecisionStrategy>
+    ) -> Result<(), ClientError> {
+
+        if !self.provider.uma2_discovered() {
+            return Err(ClientError::Uma2(NoUma2Discovered));
+        }
+
+        if self.provider.uma_policy_uri().is_none() {
+            return Err(ClientError::Uma2(NoPolicyAssociationEndpoint));
+        }
+
+        let mut url = self.provider.uma_policy_uri().unwrap().clone();
+        url.path_segments_mut()
+            .map_err(|_| ClientError::Uma2(PolicyAssociationEndpointMalformed))?
+            .extend(&[&id]);
+
+        let permission = Uma2PermissionAssociation {
+            id: Some(id),
+            name,
+            description,
+            scopes,
+            roles,
+            groups,
+            clients,
+            owner,
+            permission_type: Some("uma".to_string()),
+            logic,
+            decision_strategy
+        };
+
+        let json = self
+            .http_client
+            .put(url)
+            .header(CONTENT_TYPE, "application/json")
+            .header(AUTHORIZATION, format!("Bearer {:}", token))
```

Comment:
does it ({:}) have some special meaning?

Reply:
It is just the format! macro, unless I'm doing it wrong?

My MacBook's screen glitched up today, in for repairs. I'm out of commission with this for the next 5 to 7 business days (effectively probably next Friday) - so I still need to do the TODO items... :-/

Reply:
Take your time, we are not in the hurry...

About macro - I use either {} or {:?}, there are also some aditional variants, but I never saw {:}, so I checked documentation, have not found anything related and decided to ask you, maybe I just do not know something.

Reply:
Ok so for this mystery -
{:?} : Use Debug Trait
{:} / {} : Use Display Trait